src/http/http_client.cpp

```diff
@@ -34,6 +34,135 @@ DSN_DEFINE_uint32(http,
                   "The maximum time in milliseconds that you allow the libcurl transfer operation "
                   "to complete");
 
+#define RETURN_IF_CURL_NOT(expr, expected, ...)                                                    \
+    do {                                                                                           \
+        const auto code = (expr);                                                                  \
+        if (dsn_unlikely(code != expected)) {                                                      \
+            std::string msg(fmt::format("{}: {}", fmt::format(__VA_ARGS__), to_error_msg(code)));  \
+            return dsn::error_s::make(to_error_code(code), msg);                                   \
+        }                                                                                          \
+    } while (0)
+
+http_url::http_url() : _url(nullptr) {}
+
+http_url::~http_url() { free_curlu_object(); }
+
+http_url::http_url(http_url &&rhs) : _url(rhs._url) { rhs._url = nullptr; }
+
+http_url &http_url::operator=(http_url &&rhs)
+{
+    _url = rhs._url;
```

Comment:
Should free `_url` at first?

Then add a test to check there is no memory leak.

Reply:
I've wrapped it with `std::unique_ptr` to avoid the problem, and add test cases for copy/move constructors, copy/move assignment operators.